### PR TITLE
SURFACE_WARP and death planes use force if set

### PR DESCRIPTION
**Pull request by Reonu**
hi

### CODE REVIEW

**Comment by Reonu** [comment]
@thecozies @arthurtilly fight

File: src/game/level_update.c [hunk 1]
```diff
@@ -743,17 +743,21 @@ s16 level_trigger_warp(struct MarioState *m, s32 warpOp) {
                 break;
 
             case WARP_OP_WARP_FLOOR:
+                if (m->floor->type == SURFACE_WARP && (m->floor->force != 0x00)) {
+                    sSourceWarpNodeId = m->floor->force;
+                } else {
                 sSourceWarpNodeId = WARP_NODE_WARP_FLOOR;
```
Comment (thecozies) [inline src/game/level_update.c:749]: needs indentation

Comment (arthurtilly) [reply src/game/level_update.c:749]: indent these fucking nuts in your mouth (i agree)

**Review by arthurtilly — APPROVED** [review]
yeah yeah whatever

File: src/game/level_update.c [hunk 2]
```diff
@@ -743,17 +743,21 @@ s16 level_trigger_warp(struct MarioState *m, s32 warpOp) {
                 break;
 
             case WARP_OP_WARP_FLOOR:
-                sSourceWarpNodeId = WARP_NODE_WARP_FLOOR;
-                if (area_get_warp_node(sSourceWarpNodeId) == NULL) {
+                if (m->floor->type == SURFACE_WARP && (m->floor->force != 0x00)) {
```
Comment (gheskett) [inline src/game/level_update.c:746]: • Missing a NULL check for floor.
• The force check also needs a bitmask.

**Review by gheskett — CHANGES_REQUESTED** [review]

**Review by arthurtilly — APPROVED** [review]
still approved

File: include/config/config_collision.h [hunk 3]
```diff
@@ -17,6 +17,7 @@
 
 /**
  * Allows all surfaces types to have force, (doesn't require setting force, just allows it to be optional).
+ * Also allows you to pass a warp node to SURFACE_WARP floors via the second byte of the force parameter.
```
Comment (gheskett) [inline include/config/config_collision.h:20]: But also the death plane and vertical wind floors.

How is this actually being enforced anyway, is it pretty must just automatically handled by the nature of this define?

File: include/surface_terrains.h [hunk 4]
```diff
@@ -56,7 +56,7 @@ enum SurfaceTypes {
     SURFACE_LOOK_UP_WARP,               // 0x002F // Look up and warp (Wing cap entrance)
     SURFACE_HARD,                       // 0x0030 // Hard floor (Always has fall damage)
     SURFACE_0031,                       // 0x0031 // Unused
-    SURFACE_WARP,                       // 0x0032 // Surface warp
+    SURFACE_WARP,                       // 0x0032 // Surface warp. Warps to ID of force parameter's second byte if set, otherwise defaults to ID 0xF3.
```
Comment (gheskett) [inline include/surface_terrains.h:59]: What about vertical wind floors?

**Review by gheskett — CHANGES_REQUESTED** [review]
Comment clarifications

File: include/surface_terrains.h (hunk 4 above)
Comment (arthurtilly) [reply include/surface_terrains.h:59]: vertical wind floors have the same functionality as death planes in regards to warping

Comment (gheskett) [reply include/surface_terrains.h:59]: Exactly, but that's not mentioned anywhere here

File: include/config/config_collision.h (hunk 3 above)
Comment (arthurtilly) [reply include/config/config_collision.h:20]: idk what this means but it just checks if force isn't 0 and if a surface doesn't have force it defaults to 0

File: include/surface_terrains.h (hunk 4 above)
Comment (Reonu) [reply include/surface_terrains.h:59]: fixed

File: include/config/config_collision.h (hunk 3 above)
Comment (Reonu) [reply include/config/config_collision.h]: fixed

**Review by gheskett — APPROVED** [review]

**Pull request merged**